controller/lib/hal/hal_stm32_regs.h

```diff
@@ -0,0 +1,444 @@
+
+struct ADC_Regs {
+  // A/D specific registers (0x100 total length)
+  struct {
+    REG stat;    // 0x00 - interrupt and status register (ADC_ISR)
+    REG intEna;  // 0x04 - interrupt enable register (ADC_IER)
+    REG ctrl;    // 0x08 - control register (ADC_CR)
+    REG cfg[2];  // 0x0C - configuration registers
+    REG samp[2]; // 0x14 - sampling time registers
+    REG rsvd1;
+    REG wdog[3]; // 0x20 - watchdog threshold registers
+    REG rsvd2;
+    REG seq[4]; // 0x30 - Regular sequence registers
+    REG data;   // 0x40 - Regular data register
+    REG rsvd3[2];
+    REG iSeq; // 0x4C - Injected sequence regiseter
+    REG rsvd4[4];
+    REG offset[4]; // 0x60 - Offset registers
+    REG rsvd5[4];
+    REG iData[4]; // 0x80 - Injected channel data
+    REG rsvd6[4];
+    REG wdCfg[2]; // 0xA0 - Watchdog config
+    REG rsvd7[2];
+    REG diffSel; // 0xB0 - Differential mode selection
+    REG cal;     // 0xB4 - Calibration factors
+    REG rsvd8[18];
+  } adc[2];
+
+  // A/D common registers
+  REG comStat; // 0x300 - Common status
+  REG rsvd9;
+  REG comCtrl; // 0x304 - Common control
+  REG comData; // 0x308 - Common data
+};
+inline ADC_Regs *const ADC_BASE = reinterpret_cast<ADC_Regs *>(0X50040000);
+
+struct TimerRegs {
+  REG ctrl[2];
+  REG slaveCtrl;
+  REG intEna;
+  REG status;
+  REG event;
+  REG ccMode[2];
+  REG ccEnable;
+  REG counter;
+  REG prescale;
+  REG reload;
+  REG repeat;
+  REG compare[4];
+  REG deadTime;
+  REG dmaCtrl;
+  REG dmaAddr;
+  REG opt1;
+  REG ccMode3;
+  REG compare5;
+  REG compare6;
+  REG opt2;
+  REG opt3;
+};
+
+inline TimerRegs *const TIMER1_BASE = reinterpret_cast<TimerRegs *>(0x40012C00);
+inline TimerRegs *const TIMER2_BASE = reinterpret_cast<TimerRegs *>(0x40000000);
+inline TimerRegs *const TIMER3_BASE = reinterpret_cast<TimerRegs *>(0x40000400);
+inline TimerRegs *const TIMER6_BASE = reinterpret_cast<TimerRegs *>(0x40001000);
+inline TimerRegs *const TIMER7_BASE = reinterpret_cast<TimerRegs *>(0x40001400);
+inline TimerRegs *const TIMER15_BASE =
+    reinterpret_cast<TimerRegs *>(0x40014000);
+inline TimerRegs *const TIMER16_BASE =
+    reinterpret_cast<TimerRegs *>(0x40014400);
+
+struct FlashReg {
+  REG access;
+  REG pdKey;
+  REG key;
+  REG optKey;
+  REG status;
+  REG ctrl;
+  REG ecc;
+  REG rsvd1;
+  REG option;
+  REG pcropStart;
+  REG pcropEnd;
+  REG wrpA;
+  REG wrpB;
+};
+inline FlashReg *const FLASH_BASE = reinterpret_cast<FlashReg *>(0x40022000);
+
+struct DMA_Regs {
+  struct {
+    REG gif1 : 1;  // global interrupt flag
+    REG tcif1 : 1; // transfer complete (TC) flag
+    REG htif1 : 1; // half transfer (HT) flag
+    REG teif1 : 1; // transfer error (TE) flag
+    REG gif2 : 1;  // global interrupt flag
+    REG tcif2 : 1; // transfer complete (TC) flag
+    REG htif2 : 1; // half transfer (HT) flag
+    REG teif2 : 1; // transfer error (TE) flag
+    REG gif3 : 1;  // global interrupt flag
```

Comment:
For these DMA registers, I don't think the bit-fields improve the code.  The reason is, you have to hard code different bits for different DMA channels.  
Perhaps a better way to handle these two registers (the interrupt status & clear) would be leave them as REG values in the structure, but create an in-line function that takes an enum and the channel number and handles the access.  That would allow this to be generic for any channel, but hide the bit shifting in a function.